internal/framework/provider/config.go

```diff
@@ -0,0 +1,103 @@
+package provider
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/mongodb-forks/digest"
+	"github.com/mongodb/terraform-provider-mongodbatlas/version"
+	"github.com/spf13/cast"
+	matlasClient "go.mongodb.org/atlas/mongodbatlas"
+	realmAuth "go.mongodb.org/realm/auth"
+
+	"go.mongodb.org/realm/realm"
+)
+
+type Config struct {
+	PublicKey    string
+	PrivateKey   string
+	BaseURL      string
+	RealmBaseURL string
+	AssumeRole   *AssumeRole
+}
+
+type AssumeRole struct {
+	RoleARN           string
+	Duration          time.Duration
+	ExternalID        string
+	Policy            string
+	PolicyARNs        []string
+	SessionName       string
+	SourceIdentity    string
+	Tags              map[string]string
+	TransitiveTagKeys []string
+}
+
+// MongoDBClient client
+type MongoDBClient struct {
+	Atlas  *matlasClient.Client
+	Config *Config
+}
+
+var ua = "terraform-provider-mongodbatlas/" + version.ProviderVersion
+
+// NewClient func...
+func (c *Config) NewClient(ctx context.Context) (*MongoDBClient, error) {
+	// setup a transport to handle digest
+	transport := digest.NewTransport(cast.ToString(c.PublicKey), cast.ToString(c.PrivateKey))
+
+	// initialize the client
+	client, err := transport.Client()
+	if err != nil {
+		return nil, err
+	}
+
+	// client.Transport = logging.NewTransport("MongoDB Atlas", transport)
+
+	optsAtlas := []matlasClient.ClientOpt{matlasClient.SetUserAgent(ua)}
+	if c.BaseURL != "" {
+		optsAtlas = append(optsAtlas, matlasClient.SetBaseURL(c.BaseURL))
+	}
+
+	// Initialize the MongoDB Atlas API Client.
+	atlasClient, err := matlasClient.New(client, optsAtlas...)
+	if err != nil {
+		return nil, err
+	}
+
+	clients := &MongoDBClient{
+		Atlas:  atlasClient,
+		Config: c,
+	}
+
+	return clients, nil
+}
+
+func (c *MongoDBClient) GetRealmClient(ctx context.Context) (*realm.Client, error) {
+	// Realm
+	if c.Config.PublicKey == "" && c.Config.PrivateKey == "" {
+		return nil, errors.New("please set `public_key` and `private_key` in order to use the realm client")
+	}
```

Comment:
Should we check this also for the atlas client? if yes, can we extract the validation check into a function?